src/vm_pages.rs

```diff
@@ -482,15 +558,90 @@ impl<'a, T: GuestStagePageTable> ActiveVmPages<'a, T> {
         Ok(count)
     }
 
-    /// Returns the page fault type that corresponds to having taken `exception` at `addr`.
-    pub fn get_page_fault_cause(&self, exception: Exception, addr: GuestPhysAddr) -> PageFaultType {
-        match self.regions.find(addr) {
-            Some(VmRegionType::Confidential) => PageFaultType::Confidential(addr),
+    /// Fetches and decodes the instruction at `pc` in the guest's virtual address space.
+    fn fetch_guest_instruction(
+        &self,
+        pc: GuestVirtAddr,
+        priv_level: PrivilegeLevel,
+    ) -> Result<DecodedInstruction> {
+        // Set SPVP to reflect the privilege level we took the trap in so that
+        let old_hstatus = CSR.hstatus.get();
+        let mut hstatus = LocalRegisterCopy::<u64, hstatus::Register>::new(old_hstatus);
+        hstatus.modify(hstatus::spvp.val(priv_level as u64));
+        CSR.hstatus.set(hstatus.get());
+
+        let mut raw_inst = 0u32;
+        // Safety: _fetch_guest_instruction internally detects and handles an invalid guest virtual
+        // address in `pc'.
```

Comment:
add that it is also trusted not to write beyond the bounds of raw_inst

Reply:
Done.